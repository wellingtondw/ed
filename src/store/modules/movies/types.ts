export enum ActionTypes {
  loadingPopularMovies = 'LOADING_POPULAR_MOVIES',
  popularMoviesRequest = 'POPULAR_MOVIES_REQUEST',
  popularMoviesRequestSuccess = 'POPULAR_MOVIES_REQUEST_SUCCESS',
  popularMoviesRequestFailure = 'POPULAR_MOVIES_REQUEST_FAILURE',

  loadingSearchMovies = 'LOADING_SEARCH_MOVIES',
  searchMoviesRequest = 'SEARCH_MOVIES_REQUEST',
  searchMoviesRequestSuccess = 'SEARCH_MOVIES_REQUEST_SUCCESS',
  searchMoviesRequestFailure = 'SEARCH_MOVIES_REQUEST_FAILURE',
  resetMoviesSearch = 'RESET_MOVIES_SEARCH'
}

export type IMoviesDTO = {
  poster_path: string;
  title: string;
  release_date: string;
  vote_average: number;
  id: number;
};

export type IMoviesResponseDTO = {
  total_results: number;
  results: IMoviesDTO[];
};

type IMoviesData = {
  data: IMoviesDTO[];
  currentPage: number;
  loading: boolean;
  error: boolean;
  totalCount: number;
};

type ISearchData = {
  data: IMoviesDTO[];
  loading: boolean;
  error: boolean;
};

export type IMoviesState = {
  popularMovies: IMoviesData;
  search: ISearchData;
};
