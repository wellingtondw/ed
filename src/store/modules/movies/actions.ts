import { ActionTypes, IMoviesDTO } from './types';

//Popular Movies
export const loadingPopularMovies = () => ({
  type: ActionTypes.loadingPopularMovies
});

export const popularMoviesRequest = (page?: number) => ({
  type: ActionTypes.popularMoviesRequest,
  payload: { page }
});

type PopularMoviesRequestSuccessProps = {
  results: IMoviesDTO[];
  totalCount: number;
  page?: number;
};

export const popularMoviesRequestSuccess = ({
  results,
  totalCount,
  page
}: PopularMoviesRequestSuccessProps) => ({
  type: ActionTypes.popularMoviesRequestSuccess,
  payload: {
    results,
    totalCount,
    page
  }
});

export const popularMoviesRequestFailure = () => ({
  type: ActionTypes.popularMoviesRequestFailure
});

//Search
export const loadingSearchMovies = () => ({
  type: ActionTypes.loadingSearchMovies
});

type searchMoviesRequestProps = {
  query: string;
};

export const searchMoviesRequest = ({ query }: searchMoviesRequestProps) => ({
  type: ActionTypes.searchMoviesRequest,
  payload: { query }
});

type SearchMoviesRequestSuccessProps = {
  results: IMoviesDTO[];
};

export const searchMoviesRequestSuccess = ({ results }: SearchMoviesRequestSuccessProps) => ({
  type: ActionTypes.searchMoviesRequestSuccess,
  payload: {
    results
  }
});

export const searchMoviesRequestFailure = () => ({
  type: ActionTypes.searchMoviesRequestFailure
});

export const resetMoviesSearch = () => ({
  type: ActionTypes.resetMoviesSearch
});
