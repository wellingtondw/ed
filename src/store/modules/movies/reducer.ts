import { Reducer } from 'redux';
import { ActionTypes, IMoviesState } from './types';

export const INITIAL_STATE: IMoviesState = {
  data: {
    movies: {
      popularMovies: []
    }
  },
  loading: false,
  error: false
};

const movies: Reducer<IMoviesState> = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case ActionTypes.loading: {
      return {
        ...state,
        loading: true
      };
    }
    case ActionTypes.popularMoviesRequestSuccess: {
      return {
        ...state,
        data: {
          ...state.data,
          movies: {
            popularMovies: action.payload
          }
        },
        loading: false,
        error: false
      };
    }
    case ActionTypes.popularMoviesRequestFailure: {
      return {
        ...state,
        error: true,
        loading: false
      };
    }
    default: {
      return state;
    }
  }
};
export default movies;
