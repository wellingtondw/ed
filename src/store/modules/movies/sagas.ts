import { AxiosResponse } from 'axios';
import { all, takeLatest, call, put } from 'redux-saga/effects';
import api from '../../../services/api';

import {
  loadingPopularMovies,
  popularMoviesRequest,
  popularMoviesRequestFailure,
  popularMoviesRequestSuccess,
  loadingSearchMovies,
  searchMoviesRequest,
  searchMoviesRequestSuccess,
  searchMoviesRequestFailure
} from './actions';
import { ActionTypes, IMoviesResponseDTO } from './types';

type PopularMoviesRequest = ReturnType<typeof popularMoviesRequest>;
type SearchMoviesRequest = ReturnType<typeof searchMoviesRequest>;

const language = 'pt-BR';

export function* handleGetPopularMovies({ payload }: PopularMoviesRequest) {
  const { page } = payload;

  try {
    yield put(loadingPopularMovies());
    const { data }: AxiosResponse<IMoviesResponseDTO> = yield call(
      api.get,
      `/movie/popular?api_key=${process.env.API_KEY}&page=${page}&language=${language}`
    );

    const { results, total_results } = data;

    yield put(
      popularMoviesRequestSuccess({
        results,
        totalCount: total_results,
        page
      })
    );
  } catch (error) {
    yield put(popularMoviesRequestFailure());
  }
}

export function* handleSearchMovies({ payload }: SearchMoviesRequest) {
  const { query } = payload;

  try {
    yield put(loadingSearchMovies());
    const { data }: AxiosResponse<IMoviesResponseDTO> = yield call(
      api.get,
      `/search/movie?api_key=${process.env.API_KEY}&query=${query}&language=${language}`
    );

    const { results } = data;

    yield put(
      searchMoviesRequestSuccess({
        results
      })
    );
  } catch (error) {
    yield put(searchMoviesRequestFailure());
  }
}

export default all([
  takeLatest(ActionTypes.popularMoviesRequest, handleGetPopularMovies),
  takeLatest(ActionTypes.searchMoviesRequest, handleSearchMovies)
]);
