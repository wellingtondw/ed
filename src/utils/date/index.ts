type GetFullYearParams = {
  date: string | Date;
};
export const getFullYear = ({ date }: GetFullYearParams) => {
  return new Date(date).getFullYear();
};
