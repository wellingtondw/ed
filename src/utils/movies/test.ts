import movieCardListMock from '../../components/MovieCardList/mock';
import { formatMoviesList } from '.';

describe('utils/movies', () => {
  it('should be able to get full year from date', () => {
    const mock = movieCardListMock[0];
    const result = formatMoviesList({ items: [mock] });

    expect(result).toEqual([
      {
        id: 453395,
        date: '2022-05-04',
        posterImage: 'https://image.tmdb.org/t/p/w500/boIgXXUhw5O3oVkhXsE6SJZkmYo.jpg',
        rating: 7.5,
        title: 'Doutor Estranho no Multiverso da Loucura'
      }
    ]);
  });
});
