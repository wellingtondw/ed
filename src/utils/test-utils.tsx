import { createStore } from 'redux';
import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { createMemoryHistory } from 'history';

import { ThemeProvider } from 'styled-components';
import theme from 'styles/theme';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';

import rootReducer from '../store/modules/rootReducer';

const memoryHistory = createMemoryHistory();

const customRender = (
  ui: ReactElement,
  initialState = {},
  store = createStore(rootReducer, initialState),
  history = memoryHistory,
  { ...renderOptions }: RenderOptions = {}
) =>
  render(
    <Provider store={store}>
      <Router navigator={history} location={history.location}>
        <ThemeProvider theme={theme}>{ui}</ThemeProvider>
      </Router>
    </Provider>,
    renderOptions
  );

export * from '@testing-library/react';
export { customRender as render };
