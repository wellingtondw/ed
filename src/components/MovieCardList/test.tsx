import { render, screen } from '../../utils/test-utils';

import { formatMoviesList } from 'utils/movies';
import movieCardListMock from './mock';

import { MovieCardList } from '.';

const sut = (loading = false) => {
  return render(
    <MovieCardList loading={loading} items={formatMoviesList({ items: movieCardListMock })} />
  );
};

describe('<MovieCardList />', () => {
  it('should be able to render correctly', () => {
    const { container } = sut();

    expect(screen.getAllByRole('listitem')).toHaveLength(6);
    expect(screen.getAllByRole('img')).toHaveLength(6);
    expect(screen.queryByTestId('skeleton')).not.toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should be able to render skeleton if loading is true', () => {
    sut(true);

    expect(screen.getAllByTestId('skeleton')).toHaveLength(20);
    expect(screen.queryByRole('img')).not.toBeInTheDocument();
  });
});
