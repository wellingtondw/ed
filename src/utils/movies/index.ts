import { MovieCardListDTO } from 'components/MovieCardList';

export type MoviesDTO = {
  poster_path: string;
  title: string;
  release_date: string;
  vote_average: number;
  id: number;
};

type FormatMoviesListParams = {
  items: MoviesDTO[];
};

export const formatMoviesList = ({ items }: FormatMoviesListParams) => {
  return items.map((item): MovieCardListDTO => {
    const { poster_path, title, release_date, vote_average, id } = item;

    return {
      posterImage: poster_path,
      title,
      date: release_date,
      rating: vote_average,
      id
    };
  });
};
