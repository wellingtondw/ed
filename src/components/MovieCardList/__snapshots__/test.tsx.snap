// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCardList /> should be able to render correctly 1`] = `
.c3 {
  width: 100%;
  max-width: 250px;
  border: 1px solid #EEE;
  border-radius: 0.4rem;
  cursor: pointer;
  background-color: #FFFFFF;
  display: block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3:hover img {
  -webkit-transform: scale3d(1.05,1.05,1.05);
  -ms-transform: scale3d(1.05,1.05,1.05);
  transform: scale3d(1.05,1.05,1.05);
}

.c4 {
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  -webkit-transition: -webkit-transform 0.3s ease-in-out;
  -webkit-transition: transform 0.3s ease-in-out;
  transition: transform 0.3s ease-in-out;
  width: 100%;
}

.c6 {
  width: 100%;
  padding: 0.8rem;
}

.c7 {
  font-size: 1.6rem;
  color: #29292E;
  font-weight: 700;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-top: 4px;
}

.c9 {
  color: #835AFD;
  font-size: 1.4rem;
  font-weight: 700;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #835AFD;
  border-radius: 0.4rem;
  padding: 2px 4px;
  color: #FFFFFF;
}

.c10 svg {
  width: 1.2rem;
  height: 1.2rem;
  margin-right: 2px;
}

.c10 span {
  font-size: 1.2rem;
}

.c11 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  gap: 0.8rem;
}

.c1 {
  width: calc(20% - 16px);
  list-style: none;
}

.c1 .c2 {
  max-width: 100%;
}

@media (max-width:768px) {
  .c3 {
    max-width: 180px;
  }
}

@media (max-width:768px) {
  .c7 {
    font-size: 1.4rem;
  }
}

@media (max-width:768px) {
  .c9 {
    font-size: 1.2rem;
  }
}

@media (max-width:1170px) {
  .c1 {
    width: calc(25% - 16px);
  }
}

@media (max-width:768px) {
  .c1 {
    width: calc(33.3% - 16px);
  }
}

@media (max-width:450px) {
  .c1 {
    width: calc(50% - 16px);
  }
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <a
        class="c2 c3"
        href="/movie/453395"
      >
        <div
          class="c4"
        >
          <img
            alt="Doutor Estranho no Multiverso da Loucura"
            class="c5"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w500/boIgXXUhw5O3oVkhXsE6SJZkmYo.jpg"
          />
        </div>
        <div
          class="c6"
        >
          <h2
            class="c7"
          >
            Doutor Estranho no Multiverso da Loucura
          </h2>
          <div
            class="c8"
          >
            <span
              class="c9"
            >
              2022
            </span>
            <div
              class="c10"
            >
              <svg
                aria-hidden="true"
                class="c11"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <span>
                7.5
              </span>
            </div>
          </div>
        </div>
      </a>
    </li>
    <li
      class="c1"
    >
      <a
        class="c2 c3"
        href="/movie/4533500"
      >
        <div
          class="c4"
        >
          <img
            alt="Doutor Estranho no Multiverso da Loucura"
            class="c5"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w500/boIgXXUhw5O3oVkhXsE6SJZkmYo.jpg"
          />
        </div>
        <div
          class="c6"
        >
          <h2
            class="c7"
          >
            Doutor Estranho no Multiverso da Loucura
          </h2>
          <div
            class="c8"
          >
            <span
              class="c9"
            >
              2022
            </span>
            <div
              class="c10"
            >
              <svg
                aria-hidden="true"
                class="c11"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <span>
                7.5
              </span>
            </div>
          </div>
        </div>
      </a>
    </li>
    <li
      class="c1"
    >
      <a
        class="c2 c3"
        href="/movie/4533400"
      >
        <div
          class="c4"
        >
          <img
            alt="Doutor Estranho no Multiverso da Loucura"
            class="c5"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w500/boIgXXUhw5O3oVkhXsE6SJZkmYo.jpg"
          />
        </div>
        <div
          class="c6"
        >
          <h2
            class="c7"
          >
            Doutor Estranho no Multiverso da Loucura
          </h2>
          <div
            class="c8"
          >
            <span
              class="c9"
            >
              2022
            </span>
            <div
              class="c10"
            >
              <svg
                aria-hidden="true"
                class="c11"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <span>
                7.5
              </span>
            </div>
          </div>
        </div>
      </a>
    </li>
    <li
      class="c1"
    >
      <a
        class="c2 c3"
        href="/movie/453300"
      >
        <div
          class="c4"
        >
          <img
            alt="Doutor Estranho no Multiverso da Loucura"
            class="c5"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w500/boIgXXUhw5O3oVkhXsE6SJZkmYo.jpg"
          />
        </div>
        <div
          class="c6"
        >
          <h2
            class="c7"
          >
            Doutor Estranho no Multiverso da Loucura
          </h2>
          <div
            class="c8"
          >
            <span
              class="c9"
            >
              2022
            </span>
            <div
              class="c10"
            >
              <svg
                aria-hidden="true"
                class="c11"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <span>
                7.5
              </span>
            </div>
          </div>
        </div>
      </a>
    </li>
    <li
      class="c1"
    >
      <a
        class="c2 c3"
        href="/movie/4533200"
      >
        <div
          class="c4"
        >
          <img
            alt="Doutor Estranho no Multiverso da Loucura"
            class="c5"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w500/boIgXXUhw5O3oVkhXsE6SJZkmYo.jpg"
          />
        </div>
        <div
          class="c6"
        >
          <h2
            class="c7"
          >
            Doutor Estranho no Multiverso da Loucura
          </h2>
          <div
            class="c8"
          >
            <span
              class="c9"
            >
              2022
            </span>
            <div
              class="c10"
            >
              <svg
                aria-hidden="true"
                class="c11"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <span>
                7.5
              </span>
            </div>
          </div>
        </div>
      </a>
    </li>
    <li
      class="c1"
    >
      <a
        class="c2 c3"
        href="/movie/4533100"
      >
        <div
          class="c4"
        >
          <img
            alt="Doutor Estranho no Multiverso da Loucura"
            class="c5"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w500/boIgXXUhw5O3oVkhXsE6SJZkmYo.jpg"
          />
        </div>
        <div
          class="c6"
        >
          <h2
            class="c7"
          >
            Doutor Estranho no Multiverso da Loucura
          </h2>
          <div
            class="c8"
          >
            <span
              class="c9"
            >
              2022
            </span>
            <div
              class="c10"
            >
              <svg
                aria-hidden="true"
                class="c11"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <span>
                7.5
              </span>
            </div>
          </div>
        </div>
      </a>
    </li>
  </ul>
</div>
`;
